gio/src/cancellable.rs

```diff
@@ -1,10 +1,153 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
 use crate::prelude::*;
+use crate::Cancellable;
+use futures_channel::oneshot;
+use futures_core::Future;
+use glib::{translate::*, IsA, Source};
 
-#[test]
-fn check_callback() {
-    let c = crate::Cancellable::new();
-    c.connect_cancelled(|_| {});
-    c.cancel(); // if it doesn't crash at this point, then we're good to go!
+// rustdoc-stripper-ignore-next
+/// The id of a cancelled handler that is returned by `CancellableExtManual::connect`. This type is
+/// analogous to `glib::SignalHandlerId`.
+#[derive(Debug, Eq, PartialEq)]
+#[repr(transparent)]
+pub struct CancelledHandlerId(libc::c_ulong);
+
+impl CancelledHandlerId {
+    // rustdoc-stripper-ignore-next
+    /// Returns the internal signal handler ID.
+    pub unsafe fn as_raw(&self) -> libc::c_ulong {
+        self.0
+    }
+}
+
+impl FromGlib<libc::c_ulong> for CancelledHandlerId {
+    #[inline]
+    unsafe fn from_glib(val: libc::c_ulong) -> Self {
+        Self(val)
+    }
+}
+
+pub trait CancellableExtManual {
+    #[doc(alias = "g_cancellable_connect")]
+    fn connect<F: FnOnce(&Self) + Send + Sync + 'static>(&self, f: F) -> CancelledHandlerId;
```

Comment:
Maybe call it `connect_cancelled` / `disconnect_cancelled`? It's a better version of the existing `cancelled` signal connector